File: Vanta Speech/Vanta Speech/Core/EWS/API/EWSClient.swift
import Foundation

/// Low-level EWS SOAP client with NTLM authentication
final class EWSClient: NSObject {
    private let serverURL: URL
    private let credential: URLCredential
    private var session: URLSession!

    init(serverURL: URL, credential: URLCredential) {
        self.serverURL = serverURL
        self.credential = credential
        super.init()

        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = IntegrationConfig.EWS.requestTimeout
        config.timeoutIntervalForResource = IntegrationConfig.EWS.requestTimeout * 2

        self.session = URLSession(
            configuration: config,
            delegate: self,
            delegateQueue: nil
        )
    }

    deinit {
        session.invalidateAndCancel()
    }

    // MARK: - Public API

    /// Send a SOAP request to EWS
    /// - Parameters:
    ///   - soapAction: The SOAPAction header value
    ///   - body: The complete SOAP envelope XML
    /// - Returns: Response data
    func sendRequest(soapAction: String, body: String) async throws -> Data {
        var request = URLRequest(url: serverURL)
        request.httpMethod = "POST"
        request.httpBody = body.data(using: .utf8)
        request.setValue("text/xml; charset=utf-8", forHTTPHeaderField: "Content-Type")
        request.setValue(soapAction, forHTTPHeaderField: "SOAPAction")

        let (data, response) = try await session.data(for: request)

        guard let httpResponse = response as? HTTPURLResponse else {
            throw EWSError.networkError(URLError(.badServerResponse))
        }

        switch httpResponse.statusCode {
        case 200...299:
            return data
        case 401:
            throw EWSError.authenticationFailed
        case 403:
            throw EWSError.accessDenied
        case 429:
            throw EWSError.throttled
        case 500...599:
            // Try to extract SOAP fault
            if let fault = extractSOAPFault(from: data) {
                throw EWSError.soapFault(fault)
            }
            throw EWSError.serverError("HTTP \(httpResponse.statusCode)")
        default:
            throw EWSError.serverError("HTTP \(httpResponse.statusCode)")
        }
    }

    // MARK: - SOAP Fault Extraction

    private func extractSOAPFault(from data: Data) -> String? {
        guard let xmlString = String(data: data, encoding: .utf8) else { return nil }

        // Simple regex extraction of faultstring
        if let range = xmlString.range(of: "<faultstring>"),
           let endRange = xmlString.range(of: "</faultstring>") {
            let startIndex = range.upperBound
            let endIndex = endRange.lowerBound
            return String(xmlString[startIndex..<endIndex])
        }

        return nil
    }
}

// MARK: - URLSessionTaskDelegate (NTLM Authentication)

extension EWSClient: URLSessionTaskDelegate {
    func urlSession(
        _ session: URLSession,
        task: URLSessionTask,
        didReceive challenge: URLAuthenticationChallenge,
        completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void
    ) {
        // Prevent infinite auth loops
        guard challenge.previousFailureCount < 3 else {
            debugLog("Authentication failed after 3 attempts", module: "EWSClient", level: .error)
            completionHandler(.cancelAuthenticationChallenge, nil)
            return
        }

        let authMethod = challenge.protectionSpace.authenticationMethod

        switch authMethod {
        case NSURLAuthenticationMethodNTLM:
            debugLog("NTLM challenge received, responding with credentials", module: "EWSClient")
            completionHandler(.useCredential, credential)

        case NSURLAuthenticationMethodServerTrust:
            // Trust self-signed certs in dev (remove for production or use cert pinning)
            if let serverTrust = challenge.protectionSpace.serverTrust {
                let credential = URLCredential(trust: serverTrust)
                completionHandler(.useCredential, credential)
            } else {
                completionHandler(.performDefaultHandling, nil)
            }

        case NSURLAuthenticationMethodHTTPBasic:
            // Fallback to Basic auth if NTLM not available
            debugLog("Basic auth challenge received", module: "EWSClient")
            completionHandler(.useCredential, credential)

        default:
            debugLog("Unknown auth method: \(authMethod)", module: "EWSClient", level: .warning)
            completionHandler(.performDefaultHandling, nil)
        }
    }
}

// MARK: - Convenience Factory

extension EWSClient {
    /// Create EWSClient from stored credentials
    static func fromCredentials(_ credentials: EWSCredentials) throws -> EWSClient {
        guard let serverURL = credentials.ewsEndpoint else {
            throw EWSError.invalidServerURL
        }

        let urlCredential = URLCredential(
            user: credentials.ntlmUsername,
            password: credentials.password,
            persistence: .forSession
        )

        return EWSClient(serverURL: serverURL, credential: urlCredential)
    }
}

File: Vanta Sppech Android/app/src/main/java/com/vanta/speech/core/ews/api/EWSClient.kt
package com.vanta.speech.core.ews.api

import android.util.Log
import com.vanta.speech.core.ews.model.EWSConfig
import com.vanta.speech.core.ews.model.EWSCredentials
import com.vanta.speech.core.ews.model.EWSError
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.Authenticator
import okhttp3.Credentials
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import okhttp3.Route
import java.security.SecureRandom
import java.security.cert.X509Certificate
import java.util.concurrent.TimeUnit
import javax.net.ssl.SSLContext
import javax.net.ssl.TrustManager
import javax.net.ssl.X509TrustManager

/**
 * Low-level EWS SOAP client with NTLM/Basic authentication
 */
class EWSClient(
    private val serverURL: String,
    private val ntlmUsername: String,
    private val password: String
) {
    companion object {
        private const val TAG = "EWSClient"

        /**
         * Create EWSClient from stored credentials
         */
        fun fromCredentials(credentials: EWSCredentials): EWSClient {
            return EWSClient(
                serverURL = credentials.ewsEndpoint,
                ntlmUsername = credentials.ntlmUsername,
                password = credentials.password
            )
        }
    }

    private val client: OkHttpClient by lazy {
        createHttpClient()
    }

    private fun createHttpClient(): OkHttpClient {
        val builder = OkHttpClient.Builder()
            .connectTimeout(EWSConfig.REQUEST_TIMEOUT_MS, TimeUnit.MILLISECONDS)
            .readTimeout(EWSConfig.REQUEST_TIMEOUT_MS * 2, TimeUnit.MILLISECONDS)
            .writeTimeout(EWSConfig.REQUEST_TIMEOUT_MS, TimeUnit.MILLISECONDS)

        // Add NTLM/Basic authenticator
        builder.authenticator(object : Authenticator {
            private var authAttempts = 0

            override fun authenticate(route: Route?, response: Response): Request? {
                // Prevent infinite auth loops
                if (authAttempts >= 3) {
                    Log.e(TAG, "Authentication failed after 3 attempts")
                    authAttempts = 0
                    return null
                }
                authAttempts++

                val challenges = response.challenges()
                Log.d(TAG, "Auth challenge received: ${challenges.map { it.scheme }}")

                // Check for NTLM challenge
                val ntlmChallenge = challenges.find {
                    it.scheme.equals("NTLM", ignoreCase = true)
                }

                // Check for Basic challenge
                val basicChallenge = challenges.find {
                    it.scheme.equals("Basic", ignoreCase = true)
                }

                // Check for Negotiate (Kerberos/NTLM) challenge
                val negotiateChallenge = challenges.find {
                    it.scheme.equals("Negotiate", ignoreCase = true)
                }

                return when {
                    ntlmChallenge != null || negotiateChallenge != null -> {
                        // For NTLM, OkHttp needs NTLMEngineImpl or similar
                        // Fallback to Basic auth as most Exchange servers support both
                        Log.d(TAG, "NTLM/Negotiate detected, using Basic auth fallback")
                        response.request.newBuilder()
                            .header("Authorization", Credentials.basic(ntlmUsername, password))
                            .build()
                    }
                    basicChallenge != null -> {
                        Log.d(TAG, "Using Basic authentication")
                        response.request.newBuilder()
                            .header("Authorization", Credentials.basic(ntlmUsername, password))
                            .build()
                    }
                    else -> {
                        Log.w(TAG, "Unknown authentication challenge")
                        null
                    }
                }
            }
        })

        // Trust self-signed certs in dev (consider cert pinning for production)
        try {
            val trustAllCerts = arrayOf<TrustManager>(object : X509TrustManager {
                override fun checkClientTrusted(chain: Array<out X509Certificate>?, authType: String?) {}
                override fun checkServerTrusted(chain: Array<out X509Certificate>?, authType: String?) {}
                override fun getAcceptedIssuers(): Array<X509Certificate> = arrayOf()
            })

            val sslContext = SSLContext.getInstance("TLS")
            sslContext.init(null, trustAllCerts, SecureRandom())

            builder.sslSocketFactory(sslContext.socketFactory, trustAllCerts[0] as X509TrustManager)
            builder.hostnameVerifier { _, _ -> true }
        } catch (e: Exception) {
            Log.w(TAG, "Could not configure SSL trust: ${e.message}")
        }

        return builder.build()
    }

    /**
     * Send a SOAP request to EWS
     * @param soapAction The SOAPAction header value
     * @param body The complete SOAP envelope XML
     * @return Response data
     */
    suspend fun sendRequest(soapAction: String, body: String): ByteArray = withContext(Dispatchers.IO) {
        val mediaType = "text/xml; charset=utf-8".toMediaType()
        val requestBody = body.toRequestBody(mediaType)

        val request = Request.Builder()
            .url(serverURL)
            .post(requestBody)
            .header("Content-Type", "text/xml; charset=utf-8")
            .header("SOAPAction", soapAction)
            .build()

        val response = try {
            client.newCall(request).execute()
        } catch (e: Exception) {
            throw EWSError.NetworkError(e)
        }

        response.use { resp ->
            val responseBody = resp.body?.bytes() ?: ByteArray(0)

            when (resp.code) {
                in 200..299 -> responseBody
                401 -> throw EWSError.AuthenticationFailed
                403 -> throw EWSError.AccessDenied
                429 -> throw EWSError.Throttled
                in 500..599 -> {
                    // Try to extract SOAP fault
                    val fault = extractSOAPFault(responseBody)
                    if (fault != null) {
                        throw EWSError.SOAPFault(fault)
                    }
                    throw EWSError.ServerError("HTTP ${resp.code}")
                }
                else -> throw EWSError.ServerError("HTTP ${resp.code}")
            }
        }
    }

    private fun extractSOAPFault(data: ByteArray): String? {
        val xmlString = String(data, Charsets.UTF_8)

        // Simple regex extraction of faultstring
        val startTag = "<faultstring>"
        val endTag = "</faultstring>"

        val startIndex = xmlString.indexOf(startTag)
        val endIndex = xmlString.indexOf(endTag)

        return if (startIndex != -1 && endIndex != -1 && endIndex > startIndex) {
            xmlString.substring(startIndex + startTag.length, endIndex)
        } else {
            null
        }
    }
}
